### PR TITLE
Migrate from deprecated go-jose v2 to go-jose v3

### LINKED ISSUE #115892: Update go-jose version
### What would you like to be added?

The go-jose dependency version should be updated from v2.2.2 to v2.6.0.

### Why is this needed?

This is to make the kubernetes code compatible with newer go-jose v2 versions that have a small breaking change: `jwt.NewNumericDate()` returns a pointer in newer versions which breaks the kubernetes code [here](https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/pkg/serviceaccount/claims.go#L93):
```
/go/pkg/mod/github.com/k3s-io/kubernetes@v1.26.0-k3s1/pkg/serviceaccount/claims.go:93:29: cannot use jwt.NewNumericDate(now.Add(time.Duration(warnafter) * time.Second)) (value of type *jwt.NumericDate) as type jwt.NumericDate in assignment
```
For instance this prevents using the kubernetes apiserver and [ory/fosite](https://github.com/ory/fosite/tree/v0.44.0) together within the same Go codebase.

### LINKED ISSUE #123252: Update gopkg.in/go-jose/go-jose.v2 to v4
### What happened?

Some security scanning software is reporting a potential DoS attack in go-jose versions before 3.0.1

See: https://security.snyk.io/vuln/SNYK-GOLANG-GITHUBCOMGOJOSEGOJOSE-6070736

cc @nilekhc

### What did you expect to happen?

N/A

### How can we reproduce it (as minimally and precisely as possible)?

N/A

### Anything else we need to know?

_No response_

### Kubernetes version

N/A

### Cloud provider

N/A

### OS version

N/A

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### LINKED ISSUE #126380: Security Vulnerability: Go Package: gopkg.in/square/go-jose.v2 2.x - End of Life Notification (EOL)
Is your feature request related to a problem? Please describe.
We are currently using the rancher/hardened-kubernetes container image for our project, which was downloaded as part of the RKE2 tarball. This image internally uses version v2.6.0 of square/go-jose. Recently, we received a notification that go-jose.v2 version 2.6.0 has reached its end of life and has been deprecated. Please refer to the notification [here](https://github.com/square/go-jose/blob/master/README.md).

Describe the solution you'd like
We would like to request an upgrade to the supported version v3 of go-jose.

Additional Context:
I would like to reiterate the importance of updating the EOL packages, if a version reaches End Of Life, there won't be any security fixes done for these versions and also the vulnerabilities are not reported for these versions. This sounds like a security risk. I hope this request Is taken on priority.

### PR DESCRIPTION
**Pull request by dereknola**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Migrate from the old deprecated [go-jose/v2](https://github.com/square/go-jose?tab=readme-ov-file) to [go-jose/v3](https://github.com/go-jose/go-jose) (now operating as a independent repo not under square) 
#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/126380
Fixes https://github.com/kubernetes/kubernetes/issues/123252
Fixes https://github.com/kubernetes/kubernetes/issues/115892

#### Special notes for your reviewer:
go-jose/v2 is EOL and its last release was in 2021. The authors explicitly state that users should migrate to v3 branch.
#### Does this PR introduce a user-facing change?
NONE

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: vendor/OWNERS [hunk 1]
```diff
@@ -1,9 +0,0 @@
-# See the OWNERS docs at https://go.k8s.io/owners
```
Comment (xuzhenglun) [inline vendor/OWNERS:1]: This file should not be removed.

File: go.mod [hunk 2]
```diff
@@ -216,6 +216,7 @@ require (
 	google.golang.org/genproto/googleapis/api v0.0.0-20240528184218-531527333157 // indirect
 	gopkg.in/inf.v0 v0.9.1 // indirect
 	gopkg.in/natefinch/lumberjack.v2 v2.2.1 // indirect
+	gopkg.in/square/go-jose.v2 v2.6.0 // indirect
```
Comment (xuzhenglun) [inline go.mod:219]: It seems that there are some other dependencies is using `go-jose.v2` for now.

I did a simple grep, I may have missed something. I found that `go-jose.v2` is still used by`github.com/coreos/go-oidc` in `Verify` method.
https://github.com/kubernetes/kubernetes/blob/c22a41e879e72ba4c925b06d8aa00e43160a0f86/staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go#L645-L654

IMO, remove direct dependencies may not enough yet, those indirect dependencies have to be eliminated too.

**Comment by xuzhenglun** [comment]
Oh, I did not see paco's comment. Please ignore my review comment until you are sure that the upgrade is necessary.

**Comment by pacoxu** [comment]
/ok-to-test 
if you want to give this a try.

**Comment by dims** [comment]
also see my previous PR - https://github.com/kubernetes/kubernetes/pull/123649

there's no point wasting time if you can't get sig-auth folks time to review changes, so please speak to them first.

**Comment by dereknola** [comment]
Sorry for the quick rebases, I got confused around some of the staging files. This PR is now up-to-date with master.

**Comment by enj** [comment]
> but on a tech debt/legacy issue. There will never be a update to go-jose v2. Its dead and gone. At some point, maintaining 4+ year-old code becomes problematic when a more current, actively maintained version (go-jose v3) is available

The old versions of go-jose / go-oidc are feature complete from our perspective (we have actively shipped recent features built on them with no problems) and there are subtle behavioral changes in the newer versions of these libs that would require careful review/tweaks for us to consider upgrading.  There is no end user impact from us using the old libs, so we have not prioritized this (and have no plans to do so in the near future).  https://github.com/kubernetes/kubernetes/pull/123649#issuecomment-1975021556 sums it up well.

File: go.mod (hunk 2 above)
Comment (dereknola) [reply go.mod:219]: Moving to an indirect dependency was all I was trying to accomplish in this PR. Looking into this, we would need to do an additional migration off of `github.com/coreos/go-oidc` to `github.com/coreos/go-oidc/v3/oidc` (again a package that is from 2020 and the authors have a new branch v3 that is actually maintained).

If this PR is accepted I will do the additional work of that migration in a subsequent PR. Unless reviewers feel that they want both in this PR.

**Comment by dereknola** [comment]
Closing based on discussion. I'm guessing based of the previous number of attempt to migrate, this will not be the last PR someone opens on this topic. It may be better to leave a comment in the `go.mod` to that effect.

**Comment by akashgreninja** [comment]
Hello @enj @dims

Following up on the previous thread — I wanted to ask for clarification regarding CVE-2024-28180, which was recently flagged with a CVSS score of 4.3. for v1.32.0

From what I understand, this vulnerability still doesn't affect Kubernetes directly. Can you please confirm if that’s still the case, and whether a fix is being considered( cherry-pick ) or not?

Thanks!

**Pull request closed without merging**